awswrapper/sqs.go

```diff
@@ -0,0 +1,204 @@
+		defer close(c.StoppedChan)
+		c.Consuming = true
+		consumerTimer := time.NewTimer(c.PollTimeout)
+		for {
+			select {
+			case <-consumerTimer.C:
+				for i := 0; i < defaultBatchSize; i++ {
+					c.consume()
+				}
+				consumerTimer.Reset(c.PollTimeout)
+			case <-c.StopChan:
+				fmt.Println("stopped channel")
+				return
+			}
+		}
+	}()
+}
+
+// Stop stops a consumer from consuming messages once it completes existing payload
+func (c *Consumer) Stop() {
+	log.Println("Stopping...")
+	close(c.StopChan) // tell it to stop
+	<-c.StoppedChan   // wait for it to have stopped
+	log.Println("Stopped...")
+}
+
+// consume is a helper function for the consumer to consume messages
+func (c *Consumer) consume() {
+	// Need to convert the queue name into a URL. Make the GetQueueUrl
+	// API call to retrieve the URL. This is needed for receiving messages
+	// from the queue.
+	resultURL, err := c.SqsService.service.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: aws.String(c.QName),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == sqs.ErrCodeQueueDoesNotExist {
+			exitErrorf("Unable to find queue %q.", c.QName)
+		}
+		exitErrorf("Unable to queue %q, %v.", c.QName, err)
+	}
+
+	// Receive a message from the SQS queue with long polling enabled.
+	result, err := c.SqsService.service.ReceiveMessage(&sqs.ReceiveMessageInput{
+		QueueUrl:            resultURL.QueueUrl,
+		AttributeNames:      aws.StringSlice([]string{"SentTimestamp"}),
+		MaxNumberOfMessages: aws.Int64(c.MaxNumberOfMessagesPerRequest),
+		WaitTimeSeconds:     aws.Int64(c.WaitTimeSeconds),
+		VisibilityTimeout:   aws.Int64(c.VisibilityTimeOut),
+	})
+	if err != nil {
+		exitErrorf("Unable to receive message from queue %q, %v.", c.QName, err)
+	}
+
+	fmt.Printf("Received %d messages.\n", len(result.Messages))
+	for i, msg := range result.Messages {
+		for j := int8(1); j < c.RetryCount+1; j++ {
+			err = c.Handler(msg)
+			if err == nil {
+				c.deleteMessage(resultURL.QueueUrl, result.Messages[i])
+			} else if j == c.RetryCount {
+				fmt.Println("unable to process message. maximum retry failed.")
+				_, err = c.SqsService.SendMessage("DeadLetterQueue", result.Messages[i].String())
+				if err != nil {
+					fmt.Println("error while sending message to DLQ: ", err)
+				}
+				c.deleteMessage(resultURL.QueueUrl, result.Messages[i])
+			} else {
+				fmt.Println("consuming message failed: ", err)
+				time.Sleep(c.RetryDelay * time.Second)
+			}
+		}
+	}
+}
+
+// deleteMessage is a helper function for the consumer to delete messages
+func (c *Consumer) deleteMessage(qURL *string, msg *sqs.Message) {
+	_, err := c.SqsService.service.DeleteMessage(&sqs.DeleteMessageInput{
+		QueueUrl:      qURL,
+		ReceiptHandle: msg.ReceiptHandle,
+	})
+	if err != nil {
+		fmt.Println("Delete Error", err)
+		return
+	}
+	fmt.Println("Message Deleted")
+}
+
+func exitErrorf(msg string, args ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, msg+"\n", args...)
+	os.Exit(1)
```

Comment:
`os.Exit(1)` will kill the service using this library. We should pass the err back to the application to be handled.

Reply:
Or does it kill only the goroutine?

Reply:
I assumed the application should exit if infra has issues. Since the consumer runs in background I guess I cannot return errors to application. Will check on that.

Reply:
If it is only killing the goroutine this should be fine as long as we are logging the error. We cannot kill the application because it is serving the APIs apart from this and the request will end abruptly.